// app/elasticsearch/utils/aggParams.js
import * as _ from 'lamb';
import {
	makeMergeAppliedFnMap,
	valuesWith,
} from '@svizzle/utils';

import {
	is_required
} from 'app/elasticsearch/types/esTypes/params.utils';

import aggParamsShapeByFieldtypeByAggId from 'app/elasticsearch/config/aggParamsShapeByFieldtypeByAggId.js';
import aggParamDocByAggId from 'app/elasticsearch/config/aggParamDocByAggId';

const getParamInfo = valuesWith(
	(paramType, name) => ({
		name,
		type: paramType,
		required: is_required(paramType),
		displayText: JSON.stringify(paramType, null, 2)
	})
);

export const getTypings = _.pipe([
	getParamInfo,
	_.sortWith([_.getKey('required')])
]);

const mergeDocs = (info, docs) =>
	info.map(makeMergeAppliedFnMap({
		documentation: param => docs[param.name]
	}));

export function getParamsByAgg (aggId, fieldType) {
	const computedParams = aggParamsShapeByFieldtypeByAggId[fieldType][aggId];
	const info = getParamInfo(computedParams);
	const docs = aggParamDocByAggId[aggId];

	return mergeDocs(info, docs);
}
