// app/machines/explore/results.config.js
export const resultsConfig = {
	id: 'Results',
	initial: 'CheckMatching',
	on: {
		QUERY_UPDATED: {
			target: '#Results.CheckMatching'
		}
	},
	states: {
		CheckMatching: {
			entry: ['logEvent'],
			always: [
				{
					target: 'Matching',
					cond: 'isMatching'
				},
				{ target: 'Dirty' }
			]
		},
		Matching: {
			id: 'Matching',
			entry: [
				'logEvent',
				'sendResultsUpdated'
			],
		},
		Dirty: {
			id: 'Dirty',
			initial: 'CheckingCache',
			entry: ['logEvent'],
			states: {
				CheckingCache: {
					entry: [
						'logEvent',
						'setCacheKey',
					],
					always: [
						{
							target: '#Matching',
							cond: 'isInCache',
							actions: 'loadFromCache',
						},
						{ target: 'Pending' }
					]
				},
				Pending: {
					entry: ['logEvent'],
					invoke: {
						id: 'request',
						src: 'apiRequest',
						onDone: {
							target: '#Results.Matching',
							actions: [
								'updateCache',
								'updateCurrentResults',
							]
						},
						onError: {
							target: 'Error',
						}
					}
				},
				Error: {
					id: 'Error',
					entry: ['logEvent', 'getError'],
				}
			}
		}
	}
};
