// app/machines/explore/route.js
import {createMachina} from '../utils';

import {createHistoryStores} from './history.context';
import {historyOptions} from './history.options';
import {createResultsStores} from './results.context';
import {resultsOptions} from './results.options';
import {exploreConfig} from './route.config';
import {createRouteStores} from './route.context';
import {routeOptions} from './route.options';
import {createSelectionStores} from './selecting.context';
import {selectingOptions} from './selecting.options';

/* TODO
// nm/svizzle/utils/array-[array-object].js: makeMergeBy
// nm/app/machines/utils: mergeMachines

const mergeMachines = makeMergeBy(['actions', 'guards', 'services']);
mergeMachines([
	historyOptions,
	resultsOptions,
	routeOptions,
	selectingOptions
])
*/
export const exploreOptions = {
	actions: {
		...historyOptions.actions,
		...resultsOptions.actions,
		...routeOptions.actions,
		...selectingOptions.actions,
	},
	guards: {
		...historyOptions.guards,
		...resultsOptions.guards,
		...routeOptions.guards,
		...selectingOptions.guards,
	},
	services: {
		...resultsOptions.services,
	}
};

export const context = {
	...createHistoryStores(),
	...createResultsStores(),
	...createRouteStores(),
	...createSelectionStores(),
};

export const createExploreMachine = () => createMachina(
	exploreConfig,
	exploreOptions,
	context
);
