// app/machines/explore/search.config.js
export const searchConfig = {
	id: 'Search',
	initial: 'Idle',
	type: 'parallel',
	states: {
		SearchInput: {
			id: 'SearchInput',
			initial: 'Idle',
			states: {
				Idle: {
					entry: ['conditionalLog'],
					on: {
						TYPED: {
							target: 'Idle',
							actions: [
								'updateSearchQuery', // updates the current search query store
								'sendSearchQueryUpdated', // send('SEARCH_QUERY_UPDATED')
							]
						},
						SEARCHED: {
							target: 'Idle',
							actions: [
								'updateQuery',
								'sendQueryUpdated',
							]
						},
					},
				},
			}
		},
		Fields: {
			initial: 'Idle',
			on: {
				FIELD_SELECTED: {
					target: '.CheckingCache',
					actions: [
						'updateCurrentField',
						'sendFieldUpdated',
					]
				},
				NEXT_FIELD_SELECTED: {
					target: '.CheckingCache',
					actions: [
						'selectNextField',
						'sendFieldUpdated',
					]
				},
				PREV_FIELD_SELECTED: {
					target: '.CheckingCache',
					actions: [
						'selectPrevField',
						'sendFieldUpdated',
					]
				},
				SEARCH_QUERY_UPDATED: {
					target: '.CheckingCache',
				}
			},
			states: {
				Idle: {},
				CheckingCache: {
					entry: [
						'conditionalLog',
						'setCacheKey',
					],
					'': [
						{
							target: 'Idle',
							cond: 'isInCache',
							actions: ['updateFields'],
						},
						{ target: 'Pending' }
					]
				},
				Pending: {
					entry: ['conditionalLog'],
					invoke: {
						id: 'FieldsRequest',
						src: 'fieldsStatsRequest',
						onDone: {
							target: 'Idle',
							actions: [
								'updateCache',
								'updateFields',
							]
						},
						onError: {
							target: 'Error',
						}
					}
				},
				Error: {
					id: 'FieldsError',
					entry: [
						'conditionalLog',
						'getFieldsError'
					],
				}
			}
		},
		Suggestions: {
			initial: 'Idle',
			on: {
				FIELD_UPDATED: {
					target: '.CheckingCache',
				},
				SEARCH_QUERY_UPDATED: {
					target: '.CheckingCache'
				}
			},
			states: {
				Idle: {},
				CheckingCache: {
					entry: [
						'conditionalLog',
						'setCacheKey',
					],
					'': [
						{
							target: 'Idle',
							cond: 'isInCache',
							actions: ['updateSuggestions'],
						},
						{ target: 'Pending' }
					]
				},
				Pending: {
					entry: ['conditionalLog'],
					invoke: {
						id: 'SuggestionsRequest',
						src: 'suggestionsRequest',
						onDone: {
							target: 'Idle',
							actions: [
								'updateCache',
								'updateSuggestions',
							]
						},
						onError: {
							target: 'Error',
						}
					}
				},
				Error: {
					id: 'SuggestionsError',
					entry: [
						'conditionalLog',
						'getSuggestionsError'
					],
				}
			}
		},
	},
};
