// app/machines/explore/selecting.config.js
export const selectingConfig = {
	initial: 'Idle',
	states: {
		Idle: {
			entry: ['logState'],
			on: {
				SELECTED_FIELDS: {
					target: 'Idle',
					actions: ['logState', 'selectFields']
				},
				// ENTERED_FIELD_COUNTER: {
				// 	target: 'Idle',
				// 	actions: ['logState']
				// },
				CLICKED_FIELD_COUNTER: {
					target: 'Idle',
					actions: ['logState', 'toggleField']
				},
				// EXITED_FIELD_COUNTER: {
				// 	target: 'Idle',
				// 	actions: ['logState']
				// },
				SELECTED_NEXT_FIELD: {
					target: 'Idle',
					actions: ['logState']
				},
				SELECTED_PREVIOUS_FIELD: {
					target: 'Idle',
					actions: ['logState']
				},
				// SELECTED_NEXT_FACET: {
				// 	target: 'Idle',
				// 	actions: ['logState']
				// },
				// SELECTED_PREVIOUS_FACET: {
				// 	target: 'Idle',
				// 	actions: ['logState']
				// },
				// SELECTED_NEXT_ENTITY: {
				// 	target: 'Idle',
				// 	actions: ['logState']
				// },
				// SELECTED_PREVIOUS_ENTITY: {
				// 	target: 'Idle',
				// 	actions: ['logState']
				// },
			}
		}
	}
};
