// app/stores/exploreStores.js
import * as _ from 'lamb';
import {derived, writable} from 'svelte/store';
import {getId, mergeObj, negate} from '@svizzle/utils';

import Datasets from 'app/data/datasets.json';
import {getSchema, groupBySource} from 'app/utils/domain';
import {findValueWith} from 'app/utils/svizzle_utils/[any-boolean]-[object-any]';
// eslint-disable-next-line node/no-extraneous-import
import {makeIndexByKey} from 'svizzle/utils/array-object';
// eslint-disable-next-line node/no-extraneous-import
import {safeApply} from 'svizzle/utils/[any-any]-[any-any]';

/* sources */

const makeSources = _.pipe([
	groupBySource,
	_.mapValuesWith(_.pipe([
		_.sortWith([getId]),
		datasets => ({
			activeDatasetIndex: 0,
			datasets,
			isExpanded: false,
			isSelected: false,
		})
	]))
]);

const defaultSources = makeSources(Datasets);

export const sources = writable(defaultSources);
export const resetSources = () => {
	sources.set(defaultSources);
};
export const selectSource = sourceName => {
	sources.set(_.setPathIn(defaultSources, `${sourceName}.isSelected`, true));
};
export const toggleSource = sourceName => {
	sources.update(_.updatePath(`${sourceName}.isExpanded`, negate))
};
export const selectDataset = (sourceName, index) => {
	sources.update(_.updatePath(
		`${sourceName}`,
		mergeObj({
			isExpanded: false,
			activeDatasetIndex: index,
		}))
	)
}

/* selectedDataset */

const getIsSelected = _.getKey('isSelected');

const getSelectedDataset = _.pipe([
	findValueWith(getIsSelected),
	obj => obj && obj.datasets[obj.activeDatasetIndex]
]);
export const selectedDataset = derived(sources, getSelectedDataset);
export const selectedDatasetSchema = derived(selectedDataset, safeApply(getSchema));
export const selectedDatasetFields = derived(selectedDatasetSchema, safeApply(_.keys));
export const selectedDatasetFieldsIndicesMap = derived(selectedDatasetFields, safeApply(makeIndexByKey));
export const selectedDatasetFieldsZip = derived(selectedDatasetFields, safeApply(_.zipWithIndex));

/* navigator */

const makeNavigator = _.pipe([
	_.pairs,
	_.mapWith(([sourceName, {
		activeDatasetIndex,
		datasets,
		isExpanded,
		isSelected
	}]) => ({
		activeDataset: datasets[activeDatasetIndex],
		activeDatasetIndex,
		datasets,
		isExpanded,
		isSelected,
		sourceName,
	})),
]);

export const navigator = derived(sources, makeNavigator);
