// app/elasticsearch/config/aggsIdByFieldType.js
export default {
	boolean: [
		'cardinality',
		// 'children',
		// 'composite',
		'diversified_sampler',
		'extended_stats',
		'missing',
		'nested',
		'reverse_nested',
		'sampler',
		'value_count',
		// 'percentiles',
		// 'percentile_ranks',
	],
	date: [
		'avg',
		'cardinality',
		// 'children',
		// 'composite',
		'auto_date_histogram',
		'date_histogram',
		'date_range',
		'diversified_sampler',
		'extended_stats',
		'max',
		'median_absolute_deviation',
		'min',
		'missing',
		'nested',
		'reverse_nested',
		'percentile_ranks',
		'percentiles',
		'sampler',
		'stats',
		'sum',
		'terms',
		'value_count',
		'weighted_avg',
	],
	Date_YYYYMMDD_dash: [
		'avg',
		'cardinality',
		// 'children',
		// 'composite',
		'auto_date_histogram',
		'date_histogram',
		'date_range',
		'diversified_sampler',
		'extended_stats',
		'max',
		'median_absolute_deviation',
		'min',
		'missing',
		'nested',
		'reverse_nested',
		'percentiles',
		'sampler',
		'stats',
		'sum',
		'terms',
		'value_count',
		'weighted_avg',
	],
	Date_YYYYMMDD_dash_time: [
		'avg',
		'cardinality',
		// 'children',
		// 'composite',
		'auto_date_histogram',
		'date_histogram',
		'date_range',
		'diversified_sampler',
		'extended_stats',
		'max',
		'median_absolute_deviation',
		'min',
		'missing',
		'nested',
		'reverse_nested',
		'percentile_ranks',
		'percentiles',
		'sampler',
		'stats',
		'sum',
		'terms',
		'value_count',
		'weighted_avg',
	],
	Date_YYYYMMDD_dash_T_time: [
		'avg',
		'cardinality',
		// 'children',
		// 'composite',
		'auto_date_histogram',
		'date_histogram',
		'date_range',
		'diversified_sampler',
		'extended_stats',
		'max',
		'median_absolute_deviation',
		'min',
		'missing',
		'nested',
		'reverse_nested',
		'percentile_ranks',
		'percentiles',
		'sampler',
		'stats',
		'sum',
		'terms',
		'value_count',
		'weighted_avg',
	],
	ES_geo_point_obj: [
		'cardinality',
		// 'children',
		// 'composite',
		'diversified_sampler',
		'geo_bounds',
		'geo_centroid',
		'geo_distance',
		'geohash_grid',
		'missing',
		'nested',
		'reverse_nested',
		'sampler',
		'value_count',
	],
	ES_keyword: [
		'cardinality',
		// 'children',
		// 'composite',
		'diversified_sampler',
		'missing',
		'nested',
		'rare_terms',
		'reverse_nested',
		'sampler',
		'significant_terms',
		'significant_text',
		'terms',
		'value_count',
	],
	'ES_keyword[]': [
		'cardinality',
		// 'children',
		// 'composite',
		'diversified_sampler',
		'missing',
		'nested',
		'rare_terms',
		'reverse_nested',
		'sampler',
		'significant_terms',
		'significant_text',
		'terms',
		'value_count',
	],
	ES_text: [
		'significant_text',
	],
	'ES_text[]': [
		'significant_text',
	],
	ES_text_w_keyword: [
		'cardinality',
		// 'children',
		// 'composite',
		'diversified_sampler',
		'missing',
		'nested',
		'rare_terms',
		'reverse_nested',
		'sampler',
		'significant_terms',
		'significant_text',
		'terms',
		'value_count',
	],
	'ES_text_w_keyword[]': [
		'cardinality',
		// 'children',
		// 'composite',
		'diversified_sampler',
		'missing',
		'nested',
		'reverse_nested',
		'rare_terms',
		'sampler',
		'significant_terms',
		'significant_text',
		'text',
		'terms',
		'value_count',
	],
	float: [
		'avg',
		'cardinality',
		// 'children',
		// 'composite',
		'diversified_sampler',
		'extended_stats',
		'histogram',
		'max',
		// 'matrix_stats',
		'median_absolute_deviation',
		'min',
		'missing',
		'nested',
		'percentiles',
		'percentile_ranks',
		'reverse_nested',
		'sampler',
		'stats',
		'sum',
		'terms',
		'value_count',
		'weighted_avg',
	],
	integer: [
		'avg',
		'cardinality',
		// 'children',
		// 'composite',
		'diversified_sampler',
		'extended_stats',
		'histogram',
		'max',
		// 'matrix_stats',
		'median_absolute_deviation',
		'min',
		'missing',
		'nested',
		'percentiles',
		'percentile_ranks',
		'reverse_nested',
		'sampler',
		'stats',
		'sum',
		'terms',
		'value_count',
		'weighted_avg',
	],
	object: []
}
