// elasticsearch/types/fields.utils.js
import * as _ from 'lamb';
import {makeIsIncluded} from '@svizzle/utils';

/* ES string fields */

export const isWithKeywordTypeId =
	makeIsIncluded(['textWithKeyword', 'textWithKeywordArray']);

export const keywordFieldTypes = [
	'keyword',
	'keywordArray',
	'textWithKeyword',
	'textWithKeywordArray'
];

/* various */

export const getCountQuery = (fields, searchTerm) => ({
	size: 0,
	aggs: {
		messages: {
			filters: {
				filters: _.fromPairs(fields.map(name => [name, {
					term: {
						[name]: searchTerm
					}
				}]))
			}
		}
	}
});

export const makeGetFieldsOfTypes = types => _.pipe([
	_.pairs,
	_.filterWith(_.pipe([
		_.getPath('1.type'),
		makeIsIncluded(types)
	])),
	_.mapWith(_.getAt(0))
]);

/* dates */

/* TODO
export const isValidDate
	- isString

export const isValidDate_YYYYMMDD_dash
	- regex passes date_YYYYMMDD_dash.format.value

export const date_YYYYMMDD_dash_time
	- regex passes date_YYYYMMDD_dash_time.format.value
*/


/* Geo-point datatype */

/* TODO
export const isValidGeoPointObj
	- isShape('object')
	- lat passes numberWithin([-90, 90]),
	- lon passes numberWithin([-180, 180]),
*/
