// app/machines/builder/_context.js
// eslint-disable-next-line node/no-unpublished-import
import { writable } from 'svelte/store';

export function createBuilderStores () {
	return {
		// config
		runQueryOnSelect: writable(false),
		hideDisabledForms: writable(true),
		hideDisabledAggregations: writable(false),
		hideDisabledDatasets: writable(false),
		hideDisabledFields: writable(true),
		selectedRequestTab: writable(null),
		showFullResponse: writable(false),
		// doc strings
		activeDocs: writable(''),
		aggDocText: writable(''),
		// builder
		// forms
		forms: writable([]),
		selectedForm: writable({}),
		activeFormState: writable({}),
		// select lists
		aggregatorOptions: writable([]),
		bucketOptions: writable([]),
		typeOptions: writable([]),
		datasetOptions: writable([]),
		fieldOptions: writable([]),

		// active form
		activeParams: writable({}),
		activeSelection: writable({
			aggregation: null,
			type: null,
			dataset: null,
			field: null
		}),
		parsedQuery: writable({}),
		activeQuery: writable({}),
		resultSize: writable(0),
		readyForRequest: writable(false),
		response: writable(null),
		selectedFieldCompletions: writable([]),

		// internal stuff
		datasetTypings: null,
	}
}
