// app/machines/builder/formconfig.options.js
// eslint-disable-next-line node/no-unpublished-import
import { assign } from 'xstate';

function setTab (ctx, event) {
	ctx.selectedRequestTab.set(event.selectedRequestTab);
	return ctx
}

const toggle = v => !v;

export const configurationOptions = {
	actions: {
		/**
		 * Toggles auto-executing requests when the query is valid
		 */
		toggleAutoExecute: ctx => ctx.runQueryOnSelect.update(toggle),
		/**
		 * Hide unselectable axes. Default: true
		 */
		toggleHideDisabledForms: ctx => ctx.hideDisabledForms.update(toggle),
		/**
		 * Hide unselectable aggregations. Default: false
		 */
		toggleHideDisabledAggs: ctx => ctx.hideDisabledAggregations.update(toggle),
		/**
		 * Hide unselectable datasets. Default: false
		 */
		toggleHideDisabledDatasets: ctx => ctx.hideDisabledDatasets.update(toggle),
		/**
		 * Hide unselectable fields. Default: true
		 */
		toggleHideDisabledFields: ctx => ctx.hideDisabledFields.update(toggle),
		/**
		 * Show full response or only aggregation results.
		 */
		toggleShowFullResponse: ctx => ctx.showFullResponse.update(toggle),
		setTab: assign(setTab),

	},
	guards: {
	}
};
