// app/machines/builder/form.config.js
export const formConfig = {
	initial: 'CheckingSelection',
	onEntry: 'loadDSL',
	on: {
		SELECTION_CHANGED: {
			target: 'CheckingSelection',
			actions: [
				'clearQueryParams',
				'setSelection',
				'setURL'
			]
		},
		SELECTION_RESET: {
			target: 'SelectionIncomplete',
			actions: [
				'deleteNestedForms',
				'setURL'
			]
		},
		FORM_ACTIVATING: {
			internal: true,
			actions: [
				'activateForm',
			]
		}
	},
	states: {
		CheckingSelection: {
			onEntry: 'computeLists',
			on: {
				'': [
					{
						target: 'SelectionComplete',
						cond: 'isSelectionComplete',
						actions: [
							'spawnNestedForm'
						]
					},
					{ target: 'SelectionIncomplete' }
				]
			}
		},
		SelectionIncomplete: {
			onEntry: [
				'clearQueryParams',
				'clearTypings',
				'deleteNestedForms',
			],
		},
		SelectionComplete: {
			initial: 'CheckingQuery',
			onEntry: 'computeTypings',
			on: {
				QUERY_CHANGED: {
					target: 'SelectionComplete.CheckingQuery',
					actions: [
						'setQuery',
						'setURL'
					]
				}
			},
			states: {
				CheckingQuery: {
					id: 'CheckingQuery',
					onEntry: 'computeRequest',
					on: {
						'': [
							{
								target: 'QueryReady',
								cond: 'isQueryReady'
							},
							{ target: 'QueryNotReady' }
						]
					}
				},
				QueryNotReady: {
					id: 'QueryNotReady',
				},
				QueryReady: {
					id: 'QueryReady',
					initial: 'CheckMatching',
					states: {
						CheckMatching: {
							on: {
								'': [
									{
										target: 'Matching',
										cond: 'isMatching'
									},
									{ target: 'Dirty' }
								]
							}
						},
						Matching: {
							id: 'Matching',
							onEntry: [
								'pushHistory'
							]
						},
						Dirty: {
							id: 'Dirty',
							initial: 'Idle',
							on: {
								QUERY_EXECUTED: {
									target: 'Dirty.CheckingCache'
								}
							},
							states: {
								Idle: {
									on: {
										'': [
											{
												target: 'CheckingCache',
												cond: 'isAutoExecute'
											}
										]
									}
								},
								CheckingCache: {
									on: {
										'': [
											{
												target: '#Matching',
												cond: 'isInCache',
												actions: 'loadFromCache',
											},
											{ target: 'Pending' }
										]
									}
								},
								Pending: {
									invoke: {
										id: 'Pending',
										src: 'apiRequest',
										onDone: {
											target: '#Matching',
											actions: [
												'storeInCache',
												'setResponse'
											]
										},
										onError: {
											target: '#Error',
										}
									}
								},
								Error: {
									id: 'Error'
								}
							}
						}
					}
				}
			}
		}
	}
};
