// app/machines/builder/form.config.js
export const selectionConfig = {
	initial: 'CheckingSelection',
	on: {
		SELECTION_CHANGED: {
			target: 'Form.CheckingSelection',
			actions: [
				'computeLists',
				'setURL'
			]
		},
		SELECTION_RESET: {
			target: 'Form.SelectionIncomplete',
			actions: [
				'deleteNestedForms',
				'setURL'
			]
		}
	},
	states: {
		CheckingSelection: {
			always: [
				{
					target: 'SelectionComplete',
					cond: 'isSelectionComplete',
					actions: [
						'prepareForm',
						'spawnNestedForm'
					]
				},
				{ target: 'SelectionIncomplete' }
			]
		},
		SelectionIncomplete: {},
		SelectionComplete: {
			initial: 'CheckingQuery',
			on: {
				QUERY_CHANGED: {
					target: 'SelectionComplete.CheckingQuery',
					actions: [
						'computeRequest',
						'setURL'
					]
				}
			},
			states: {
				CheckingQuery: {
					id: 'CheckingQuery',
					always: [
						{
							target: 'QueryReady',
							cond: 'isQueryReady'
						},
						{ target: 'QueryNotReady' }
					]
				},
				QueryNotReady: {
					id: 'QueryNotReady',
				},
				QueryReady: {
					id: 'QueryReady',
					initial: 'CheckMatching',
					states: {
						CheckMatching: {
							always: [
								{
									target: 'Matching',
									cond: 'isMatching'
								},
								{ target: 'Dirty' }
							]
						},
						Matching: {
							id: 'Matching',
							onEntry: [
								'pushHistory'
							]
						},
						Dirty: {
							id: 'Dirty',
							initial: 'Idle',
							on: {
								QUERY_EXECUTED: {
									target: 'Dirty.CheckingCache'
								}
							},
							states: {
								Idle: {
									always: [
										{
											target: 'CheckingCache',
											cond: 'isAutoExecute'
										}
									]
								},
								CheckingCache: {
									onEntry: ['searchInCache'],
									always: [
										{
											target: '#Matching',
											cond: 'isInCache'
										},
										{ target: 'Pending' }
									],
								},
								Pending: {
									invoke: {
										id: 'Pending',
										src: 'apiRequest',
										onDone: {
											target: '#Matching',
											actions: ['storeInCache']
										},
										onError: {
											target: '#Error',
										}
									}
								},
								Error: {
									id: 'Error'
								}
							}
						}
					}
				}
			}
		}
	}
};
