// app/machines/builder/route.js
// eslint-disable-next-line node/no-unpublished-import
import { Machine } from 'xstate';

import { configurationConfig } from './formconfig.config';
import { configurationOptions } from './formconfig.options';
import { selectionConfig } from './form.config';
import { selectionOptions } from './form.options';
import { docsConfig } from './docs.config';

export const routeConfig = {
	id: 'Route',
	initial: 'Loading',
	on: {
		ROUTE_CHANGED: {
			target: 'Navigating'
		}
	},
	states: {
		Loading: {
			on: {
				READY: {
					target: 'Interactive',
					actions: ['parseParams']
				}
			}
		},
		Navigating: {
			on: {
				NAVIGATED: {
					target: 'Interactive',
					actions: ['parseParams']
				}
			}
		},
		Interactive: {
			type: 'parallel',
			states: {
				FormConfig: configurationConfig,
				Docs: docsConfig,
				Form: selectionConfig,
			}
		}
	}
};

export const routeOptions = {
	actions: {
		...configurationOptions.actions,
		...selectionOptions.actions
	},
	guards: {
		...configurationOptions.guards,
		...selectionOptions.guards
	}
};

export const RouteMachine = Machine(routeConfig, routeOptions);
