// app/elasticsearch.js
export const descriptionsEN = {
    avg: "Average",
    weighted_avg: "Weighted Average",
    cardinality: "Cardinality",
    extended_stats: "Extended Statistics",
    geo_bounds: "Geographic Bounds",
    geo_centroid: "Geographic Centroid",
    max: "Maximum",
    min: "Minimum",
    percentiles: "Percentiles",
    percentile_ranks: "Percentile Ranks",
    scripted_metric: "Scripted Metric",
    stats: "Statistics",
    string_stats: "String Statistics",
    sum: "Sum",
    top_tags: "Top Hits",
    types_count: "Value Count",
    median_absolute_deviation: "Median Absolute Deviation"
}

export const aggregationsPerType = {
    geo_point: [
        'cardinality', 
        'geo_bounds', 
        'geo_centroid', 
        //'types_count'
    ],
    boolean: [
        'cardinality', 
        //'percentile_ranks', 
        //'types_count'
    ],
    integer: [
        //'percentile_ranks', 
        'extended_stats', 
        //'percentile_ranks', 
        //'types_count'
    ],
    Date_YYYYMMDD_dash: [
        'cardinality', 
        'extended_stats', 
        'percentiles', 
        //'percentile_ranks', 
        //'types_count'
    ],
    date: [
        'cardinality', 
        'percentiles', 
        //'percentile_ranks', 
        //'types_count'
    ],
    string: [
        //'string_stats'
    ],
    float: [
        //'weighted_avg', 
        'cardinality', 
        'extended_stats', 
        'percentiles', 
        //'percentile_ranks', 
        //'types_count'
    ],
    'string[]': [
        //'types_count'
    ],
    object: []
}

export function determineESType(field) {
    if (typeof field === 'string')
        return field;
    if (typeof field === 'object') {
        // TODO: We should convene on one of these. I'm adding both for now.
        if ('es_type' in field)
            return field.es_type;
        if ('type' in field)
            return field.type;

        // TODO: Should we throw or simply return 'object'?
        // throw new Error('Bad type specification');
    }
    return 'object';
}

const blah = {
    float: {

    }
}
