// app/elasticsearch/config.js
export const metricDescriptionsEN = {
	// median_absolute_deviation: "Median Absolute Deviation"
	// percentile_ranks: "Percentile Ranks",
	// scripted_metric: "Scripted Metric",
	// string_stats: "String Statistics",
	// top_hits: "Top Hits",
	// weighted_avg: "Weighted Average",
	avg: "Average",
	cardinality: "Cardinality",
	extended_stats: "Extended Statistics",
	geo_bounds: "Geographic Bounds",
	geo_centroid: "Geographic Centroid",
	max: "Maximum",
	min: "Minimum",
	percentiles: "Percentiles",
	stats: "Statistics",
	sum: "Sum",
	value_count: "Value Count",
}

export const bucketDescriptionsEN = {
	histogram: "Histogram",
	date_histogram: "Date Histogram",
	missing: "Missing",
	terms: "Terms"
}

export const aggregationsPerType = {
	boolean: [
		'avg',
		'cardinality',
		'max',
		'min',
		'missing',
		'stats',
		'sum',
		'value_count',
		// 'percentile_ranks',
	],
	date: [
		'avg',
		'cardinality',
		'date_histogram',
		'extended_stats',
		'max',
		'min',
		'missing',
		'percentiles',
		'stats',
		'sum',
		'terms',
		'value_count',
		// 'percentile_ranks',
	],
	Date_YYYYMMDD_dash: [
		'avg',
		'cardinality',
		'date_histogram',
		'extended_stats',
		'max',
		'min',
		'missing',
		'percentiles',
		'stats',
		'sum',
		'terms',
		'value_count',
		// 'percentile_ranks',
	],
	Date_YYYYMMDD_dash_time: [
		'avg',
		'cardinality',
		'date_histogram',
		'extended_stats',
		'max',
		'min',
		'missing',
		'percentiles',
		'stats',
		'sum',
		'terms',
		'value_count',
		// 'percentile_ranks',
	],
	Date_YYYYMMDD_dash_T_time: [
		'avg',
		'cardinality',
		'date_histogram',
		'extended_stats',
		'max',
		'min',
		'missing',
		'percentiles',
		'stats',
		'sum',
		'terms',
		'value_count',
		// 'percentile_ranks',
	],
	ES_geo_point_obj: [
		'cardinality',
		'geo_bounds',
		'geo_centroid',
		'missing',
		'value_count',
	],
	ES_keyword: [
		'cardinality',
		'terms',
		'value_count',
		// 'string_stats' //disabled until backend upgrade
	],
	'ES_keyword[]': [
		'cardinality',
		'terms',
		'value_count',
		// 'string_stats' //disabled until backend upgrade
	],
	ES_text_w_keyword: [
		'cardinality',
		'terms',
		'value_count',
		// 'string_stats' //disabled until backend upgrade
	],
	'ES_text_w_keyword[]': [
		'cardinality',
		'terms',
		'value_count',
		// 'string_stats' //disabled until backend upgrade
	],
	float: [
		'avg',
		'cardinality',
		'extended_stats',
		'histogram',
		'max',
		'min',
		'missing',
		'percentiles',
		'stats',
		'sum',
		'value_count',
		// 'percentile_ranks',
		// 'weighted_avg',
	],
	integer: [
		'avg',
		'cardinality',
		'extended_stats',
		'histogram',
		'max',
		'min',
		'missing',
		'percentiles',
		'stats',
		'sum',
		'terms',
		'value_count',
		// 'percentile_ranks',
	],
	object: []
}
