// app/elasticsearch/index.js
import * as _ from 'lamb';
import {
	isObject,
	isString,
	makeMergeAppliedFnMap,
} from '@svizzle/utils';

import { evaluate, getParamInfo } from './utils';
import {
	aggsByType,
	defaultValues,
	docsByParamByAgg,
	paramsByAggs
} from './config';

const ifExistsGet = prop => _.casus(_.hasKey(prop), _.getKey(prop));

export const getESType = _.adapter([
	_.casus(isString, _.identity),
	_.casus(isObject, _.adapter([
		ifExistsGet('es_type'),
		ifExistsGet('data_type'),
		ifExistsGet('type'),
	])),
	_.always('object')
]);

const mergeDocs = (info, docs) =>
	info.map(makeMergeAppliedFnMap({
		documentation: param => docs[param.name]
	}));

export function getParamsByAgg (agg, type) {
	const params = paramsByAggs[agg];
	const computedParams = evaluate(type)(params);
	const info = getParamInfo(computedParams);
	const docs = docsByParamByAgg[agg];
	return mergeDocs(info, docs);
}

const hasKeywordField = _.anyOf([
	_.is('ES_text_w_keyword'),
	_.is('ES_text_w_keyword[]'),
]);

const createKeywordFieldQuery = fieldName => ({field: `${fieldName}.keyword`});

function defaultAggregationFactory (fieldName, fieldInfo) {
	const fieldType = getESType(fieldInfo);
	if (hasKeywordField(fieldType)) {
		return createKeywordFieldQuery(fieldName);
	}
	return { field: fieldName };
}

const queryFactories = {
	composite: () => ({}),
	filter: () => ({}),
	filters: () => ({}),
	sampler: () => ({}),
	// eslint-disable-next-line no-unused-vars
	weighted_avg: () => ({})
}

export function buildAggregation (aggregation, fieldName, fieldInfo) {
	if (aggregation in queryFactories) {
		return queryFactories[aggregation](fieldName, fieldInfo);
	}

	return defaultAggregationFactory(fieldName, fieldInfo);
}

export function constructQuery (schema) {
	const aggs = {};
	for (let f in schema) {
		const fType = getESType(schema[f]);
		const typeAggs = aggsByType[fType];
		for (let i in typeAggs) {
			const at = typeAggs[i];
			const atName = `${f}_${at}`;
			aggs[atName] = {
				[at]: buildAggregation(at, f, schema[f])
			};
		}
	}

	return {
		size: 0,
		aggs
	};
}

export function getDefaultValue (aggregation, dataset, field, type, parameter) {
	if (!(aggregation in defaultValues)) {
		return undefined;
	}
	const agg = defaultValues[aggregation];
	if (!(parameter in agg)) {
		return undefined;
	}
	const defaultValue = agg[parameter]
	if (typeof defaultValue !== 'function') {
		return defaultValue;
	}
	const computedValue = defaultValue(dataset, field, type);
	if (computedValue && computedValue.then && typeof computedValue.then === 'function') {
		// TODO wait until promise resolution
	} else {
		return computedValue;
	}
	return undefined;
}

export const getAggDocs = agg => docsByParamByAgg[agg].__docs;
