// app/tsservices/testsrc.js
export const src = `
// Utility types

type ExactlyOneOfSome<T, K extends keyof T = keyof T> = K extends unknown ?
    { 
        [I in keyof T]?: I extends K ?
            T[I] :
            never;
    } : 
    never;

type ExactlyOne<T> = ExactlyOneOfSome<T, keyof T>;

// ElasticSearch Typings
// ElasticSearch Helper Types

export type ScriptLanguage = 'painless' | 'expression' | 'mustache' | 'java';

export type SortOrder = 'asc' | 'desc';

export type SortInstruction<Dataset> = 
    Record<keyof Dataset, SortOrder | { order: SortOrder }>;

export type SortOptions<Dataset> = 
    SortOrder | 
    SortInstruction<Dataset> | 
    SortInstruction<Dataset>[];

export type SourceOptions<Dataset> = string | string[];  //TODO incomplete


//Aggregation Helper Types

// This is a rather 'abstract' base interface
export type ScriptBase = {
	lang?: ScriptLanguage;
	params?: Record<string, string | number>;
}

export type LiteralScript = ScriptBase | {
	source: string;
}

export type ScriptReference = ScriptBase | {
	id: string;
}

// These are the actual values accepted by ES.
// Note that ScriptBase is not icluded
export type Script = {
    script: string | LiteralScript | ScriptReference;
}

export type MetricField<Dataset, Key extends keyof Dataset = keyof Dataset> = 
    Key extends keyof Dataset ?
        {
            field: Key;
            missing?: Dataset[Key];
        } :
            never;

export type MetricAggregationConfig<Dataset, Keys extends keyof Dataset> = 
    MetricField<Dataset, Keys> | Script;


export type MetricAggregations<Dataset, Fields extends keyof Dataset> = {
    avg: MetricAggregationConfig<Dataset, Fields>;
    cardinality: MetricAggregationConfig<Dataset, Fields> & {
        precision_threshold?: number;
    };
    extended_stats: MetricAggregationConfig<Dataset, Fields>;
    geo_bounds: MetricAggregationConfig<Dataset, Fields> & {
        wrap_longitude?: boolean;
    };
    geo_centroid: MetricAggregationConfig<Dataset, Fields>;
    max: MetricAggregationConfig<Dataset, Fields>;
    min: MetricAggregationConfig<Dataset, Fields>;
    percentile_ranks: MetricAggregationConfig<Dataset, Fields> & {
    };
    percentiles: MetricAggregationConfig<Dataset, Fields> & {
        percents?: number[];
		hdr?: { 
            number_of_significant_value_digits: number;
        };
    };
    scripted_metric: {
        init_script: Script;
        map_script: Script;
        combine_script: Script;
        reduce_script: Script;
    }
    stats: MetricAggregationConfig<Dataset, Fields>;
    string_stats: MetricAggregationConfig<Dataset, Fields> & {
        show_distribution?: boolean;
    };
    sum: MetricAggregationConfig<Dataset, Fields>;
    top_hits: {
		from?: number;
		size?: number;
		sort?: SortOptions<Dataset>;
		_source?: SourceOptions<Dataset>;
    };
    value_count: MetricAggregationConfig<Dataset, Fields>;
    weighted_avg: MetricAggregationConfig<Dataset, Fields> & {
        values: Dataset[Fields][];
    };
}

export type BucketAggregations<Dataset, Fields extends keyof Dataset> = {
    date_histogram: MetricAggregationConfig<Dataset, Fields> & {
        interval: '1m' | '1h' | '1d' | '1w' | '1M' | '1q' | '1y';  // ES < 7
        //calendar_interval: '1m' | '1h' | '1d' | '1w' | '1M' | '1q' | '1y'; // ES >= 7
        //fixed_interval: '1m' | '1h' | '1d' | '1w' | '1M' | '1q' | '1y'; // ES >= 7 and also needs more thought
    };
    histogram: MetricAggregationConfig<Dataset, Fields> & {
        interval: number;
    };
    missing:  MetricField<Dataset, Fields>;
    terms: MetricAggregationConfig<Dataset, Fields> & {

    };
}

export type MatrixAggregations<Dataset, Fields extends keyof Dataset> = {
    matrix_stats: {
        fields: Fields[]
        missing?: {
            [K in Fields]: Dataset[K];
        }
    }
}

export type Aggregations<Dataset, Fields extends keyof Dataset> =
    MetricAggregations<Dataset, Fields> &
    BucketAggregations<Dataset, Fields> & 
    MatrixAggregations<Dataset, Fields>;

type Aggregation<Dataset, DatasetField extends keyof Dataset> =
	ExactlyOne<Aggregations<Dataset, DatasetField>>;

type Aggs<Dataset, DatasetField extends keyof Dataset> = {
    [aggName: string]: Aggregation<Dataset, DatasetField> & {
        aggs?: Aggs<Dataset, DatasetField>;
    };
};
export type Opaque<T, K> = T & { __opaque__: K };

export type date = Opaque<Date, 'date'>;
export type float = Opaque<number, 'float'>;
export type integer = Opaque<number, 'integer'>;

export type Date_YYYYMMDD = Opaque<string, 'Date_YYYYMMDD'>;
export type Date_YYYYMMDD_dash = Opaque<string, 'Date_YYYYMMDD_dash'>;
export type Date_YYYYMMDD_dash_time = Opaque<string, 'Date_YYYYMMDD_dash_time'>;
export type Date_YYYYMMDD_dash_T_time = Opaque<string, 'Date_YYYYMMDD_dash_T_time'>;

export type ES_geo_point_obj = {
    lat: float,
    lon: float
}

export type ES_keyword = Opaque<string, 'ES_keyword'>;
export type ES_text = Opaque<string, 'ES_text'>;
export type ES_text_array = Opaque<string, 'ES_text_array'>;
export type ES_text_w_keyword  = Opaque<string, 'ES_text_w_keyword'>;
export type arxlive_arxiv_v3 = {
	booleanFlag_multinational_article: boolean;
	count_citations_article: integer;
	date_created_article: date;
	id_digitalObjectIdentifier_article: ES_text;
	json_category_article: object[];
	json_fieldOfStudy_article: object[];
	json_location_article: object[];
	metric_citations_article: float;
	metric_novelty_article: float;
	terms_authors_article: ES_text_w_keyword[];
	terms_institutes_article: ES_text_w_keyword[];
	terms_tokens_article: ES_keyword[];
	textBody_abstract_article: ES_text;
	title_of_article: ES_text;
	type_of_entity: ES_text;
	url_of_article: ES_text;
	year_of_article: integer;
};

export type arxlive_arxiv_v4 = {
	booleanFlag_multinational_article: boolean;
	count_citations_article: integer;
	date_created_article: date;
	id_digitalObjectIdentifier_article: ES_keyword;
	json_category_article: object[];
	json_fieldOfStudy_article: object[];
	json_location_article: object[];
	metric_citations_article: float;
	metric_novelty_article: float;
	terms_authors_article: ES_text_w_keyword[];
	terms_institutes_article: ES_text_w_keyword[];
	terms_tokens_article: ES_keyword[];
	textBody_abstract_article: ES_text_w_keyword;
	title_of_article: ES_text_w_keyword;
	type_of_entity: ES_text_w_keyword;
	url_of_article: ES_text_w_keyword;
	year_of_article: integer;
};

export type eurito_arxiv_v0 = {
	booleanFlag_eu_article: boolean;
	booleanFlag_multinational_article: boolean;
	count_citations_article: integer;
	date_created_article: Date_YYYYMMDD_dash;
	id_digitalObjectIdentifier_article: ES_keyword;
	json_fieldsOfStudy_article: object;
	metric_novelty_article: float;
	terms_authors_article: ES_text_w_keyword[];
	terms_category_article: ES_text_w_keyword[];
	terms_countries_article: ES_text_w_keyword[];
	terms_fieldsOfStudy_article: ES_text_w_keyword[];
	terms_institutes_article: ES_text_w_keyword[];
	terms_nuts0_article: ES_text_w_keyword[];
	terms_nuts1_article: ES_text_w_keyword[];
	terms_nuts2_article: ES_text_w_keyword[];
	terms_nuts3_article: ES_text_w_keyword[];
	terms_regions_article: ES_text_w_keyword[];
	terms_tokens_entity: ES_keyword[];
	textBody_abstract_article: ES_text;
	title_of_article: ES_text;
	type_of_entity: ES_keyword;
	url_of_article: ES_text;
	year_of_article: integer;
};

export type eurito_arxiv_v1 = {
	booleanFlag_eu_article: boolean;
	booleanFlag_multinational_article: boolean;
	count_citations_article: integer;
	date_created_article: Date_YYYYMMDD_dash;
	id_digitalObjectIdentifier_article: ES_keyword;
	json_fieldsOfStudy_article: object;
	metric_novelty_article: float;
	terms_authors_article: ES_text_w_keyword[];
	terms_category_article: ES_text_w_keyword[];
	terms_countries_article: ES_text_w_keyword[];
	terms_fieldsOfStudy_article: ES_text_w_keyword[];
	terms_institutes_article: ES_text_w_keyword[];
	terms_nuts0_article: ES_text_w_keyword[];
	terms_nuts1_article: ES_text_w_keyword[];
	terms_nuts2_article: ES_text_w_keyword[];
	terms_nuts3_article: ES_text_w_keyword[];
	terms_regions_article: ES_text_w_keyword[];
	terms_tokens_entity: ES_keyword[];
	textBody_abstract_article: ES_text_w_keyword;
	title_of_article: ES_text_w_keyword;
	type_of_entity: ES_keyword;
	url_of_article: ES_text_w_keyword;
	year_of_article: integer;
};

export type eurito_cordis_v0 = {
	cost_ecFunding_project: integer;
	cost_total_project: integer;
	date_ended_project: Date_YYYYMMDD_dash;
	date_started_project: Date_YYYYMMDD_dash;
	metric_novelty_project: float;
	name_framework_project: ES_text;
	status_of_project: ES_keyword;
	terms_tokens_entity: ES_keyword[];
	textBody_description_project: ES_text;
	title_of_project: ES_text;
	type_of_entity: ES_keyword;
	url_of_project: ES_text;
	year_of_project: integer;
};

export type eurito_cordis_v1 = {
	cost_ecFunding_project: integer;
	cost_total_project: integer;
	date_ended_project: Date_YYYYMMDD_dash;
	date_started_project: Date_YYYYMMDD_dash;
	metric_novelty_project: float;
	name_framework_project: ES_text_w_keyword;
	status_of_project: ES_keyword;
	terms_tokens_entity: ES_keyword[];
	textBody_description_project: ES_text_w_keyword;
	title_of_project: ES_text_w_keyword;
	type_of_entity: ES_keyword;
	url_of_project: ES_text_w_keyword;
	year_of_project: integer;
};

export type eurito_crunchbase_v0 = {
	_cost_usd2018_organisation: float;
	_terms_sdg_summary: ES_keyword[];
	address_of_organisation: ES_keyword;
	booleanFlag_eu_organisation: boolean;
	booleanFlag_health_organisation: boolean;
	coordinate_of_city: ES_geo_point_obj;
	cost_of_funding: float;
	count_employee_organisation: ES_keyword;
	count_rounds_funding: integer;
	currency_of_funding: ES_keyword;
	date_birth_organisation: Date_YYYYMMDD_dash;
	date_death_organisation: Date_YYYYMMDD_dash;
	date_last_funding: Date_YYYYMMDD_dash;
	date_updated_organisation: Date_YYYYMMDD_dash;
	id_continent_organisation: ES_keyword;
	id_iso2_country: ES_keyword;
	id_iso3_country: ES_keyword;
	id_isoNumeric_country: integer;
	id_of_continent: ES_keyword;
	id_parent_organisation: ES_keyword;
	id_state_organisation: ES_keyword;
	metric_novelty_organisation: float;
	name_of_organisation: ES_text;
	placeName_city_organisation: ES_text_w_keyword;
	placeName_continent_organisation: ES_text_w_keyword;
	placeName_country_organisation: ES_text_w_keyword;
	placeName_region_organisation: ES_text_w_keyword;
	placeName_state_organisation: ES_text_w_keyword;
	rank_rhodonite_organisation: float;
	status_of_organisation: ES_keyword;
	terms_alias_organisation: ES_keyword[];
	terms_category_organisation: ES_keyword[];
	terms_mesh_description: ES_text_w_keyword[];
	terms_of_countryTags: ES_keyword[];
	terms_of_funders: ES_text_w_keyword[];
	terms_roles_organisation: ES_text_w_keyword[];
	terms_subcategory_organisation: ES_text_w_keyword[];
	terms_tokens_entity: ES_keyword[];
	textBody_descriptive_organisation: ES_text;
	textBody_summary_organisation: ES_text;
	type_of_entity: ES_keyword;
	type_of_organisation: ES_keyword;
	url_crunchBase_organisation: ES_keyword;
	url_facebook_organisation: ES_keyword;
	url_linkedIn_organisation: ES_keyword;
	url_of_organisation: ES_keyword;
	url_twitter_organisation: ES_keyword;
};

export type eurito_crunchbase_v1 = {
	_cost_usd2018_organisation: float;
	_terms_sdg_summary: ES_keyword[];
	address_of_organisation: ES_keyword;
	booleanFlag_eu_organisation: boolean;
	booleanFlag_health_organisation: boolean;
	coordinate_of_city: ES_geo_point_obj;
	cost_of_funding: float;
	count_employee_organisation: ES_keyword;
	count_rounds_funding: integer;
	currency_of_funding: ES_keyword;
	date_birth_organisation: Date_YYYYMMDD_dash;
	date_death_organisation: Date_YYYYMMDD_dash;
	date_last_funding: Date_YYYYMMDD_dash;
	date_updated_organisation: Date_YYYYMMDD_dash;
	id_continent_organisation: ES_keyword;
	id_iso2_country: ES_keyword;
	id_iso3_country: ES_keyword;
	id_isoNumeric_country: integer;
	id_of_continent: ES_keyword;
	id_parent_organisation: ES_keyword;
	id_state_organisation: ES_keyword;
	metric_novelty_organisation: float;
	name_of_organisation: ES_text_w_keyword;
	placeName_city_organisation: ES_text_w_keyword;
	placeName_continent_organisation: ES_text_w_keyword;
	placeName_country_organisation: ES_text_w_keyword;
	placeName_region_organisation: ES_text_w_keyword;
	placeName_state_organisation: ES_text_w_keyword;
	rank_rhodonite_organisation: float;
	status_of_organisation: ES_keyword;
	terms_alias_organisation: ES_keyword[];
	terms_category_organisation: ES_keyword[];
	terms_mesh_description: ES_text_w_keyword[];
	terms_of_countryTags: ES_keyword[];
	terms_of_funders: ES_text_w_keyword[];
	terms_roles_organisation: ES_text_w_keyword[];
	terms_subcategory_organisation: ES_text_w_keyword[];
	terms_tokens_entity: ES_keyword[];
	textBody_descriptive_organisation: ES_text_w_keyword;
	textBody_summary_organisation: ES_text_w_keyword;
	type_of_entity: ES_keyword;
	type_of_organisation: ES_keyword;
	url_crunchBase_organisation: ES_text_w_keyword;
	url_facebook_organisation: ES_text_w_keyword;
	url_linkedIn_organisation: ES_text_w_keyword;
	url_of_organisation: ES_text_w_keyword;
	url_twitter_organisation: ES_text_w_keyword;
};

export type eurito_patstat_v0 = {
	booleanFlag_autotranslated_entity: boolean;
	booleanFlag_eu_patent: boolean;
	count_citations_patent: integer;
	date_of_patent: Date_YYYYMMDD_dash;
	metric_novelty_patent: float;
	terms_authCountry_patent: ES_text_w_keyword[];
	terms_ipc_patent: ES_text_w_keyword[];
	terms_iso2lang_entity: ES_text_w_keyword[];
	terms_nace2_patent: ES_text_w_keyword[];
	terms_personCountry_patent: ES_text_w_keyword[];
	terms_personNuts_patent: ES_text_w_keyword[];
	terms_techFieldNumber_patent: ES_text_w_keyword[];
	terms_tokens_entity: ES_keyword[];
	textBody_abstract_patent: ES_text;
	title_of_patent: ES_text;
	type_of_entity: ES_keyword;
	year_of_patent: integer;
};

export type eurito_patstat_v1 = {
	booleanFlag_autotranslated_entity: boolean;
	booleanFlag_eu_patent: boolean;
	count_citations_patent: integer;
	date_of_patent: Date_YYYYMMDD_dash;
	metric_novelty_patent: float;
	terms_authCountry_patent: ES_text_w_keyword[];
	terms_ipc_patent: ES_text_w_keyword[];
	terms_iso2lang_entity: ES_text_w_keyword[];
	terms_nace2_patent: ES_text_w_keyword[];
	terms_personCountry_patent: ES_text_w_keyword[];
	terms_personNuts_patent: ES_text_w_keyword[];
	terms_techFieldNumber_patent: ES_text_w_keyword[];
	terms_tokens_entity: ES_keyword[];
	textBody_abstract_patent: ES_text_w_keyword;
	title_of_patent: ES_text_w_keyword;
	type_of_entity: ES_keyword;
	year_of_patent: integer;
};

export type mosaic_crunchbase_v4 = {
	_cost_usd2018_organisation: float;
	_terms_sdg_summary: ES_keyword[];
	address_of_organisation: ES_keyword;
	booleanFlag_health_organisation: boolean;
	coordinate_of_city: ES_geo_point_obj;
	cost_of_funding: float;
	count_employee_organisation: ES_keyword;
	count_rounds_funding: integer;
	currency_of_funding: ES_keyword;
	date_birth_organisation: Date_YYYYMMDD_dash;
	date_death_organisation: Date_YYYYMMDD_dash;
	date_last_funding: Date_YYYYMMDD_dash;
	datetime_updated_organisation: Date_YYYYMMDD_dash_time;
	id_continent_organisation: ES_keyword;
	id_iso2_country: ES_keyword;
	id_iso3_country: ES_keyword;
	id_isoNumeric_country: integer;
	id_of_continent: ES_keyword;
	id_parent_organisation: ES_keyword;
	id_state_organisation: ES_keyword;
	name_of_organisation: ES_text;
	placeName_city_organisation: ES_text_w_keyword;
	placeName_continent_organisation: ES_text_w_keyword;
	placeName_country_organisation: ES_text_w_keyword;
	placeName_region_organisation: ES_text_w_keyword;
	placeName_state_organisation: ES_text_w_keyword;
	rank_rhodonite_organisation: float;
	status_of_organisation: ES_keyword;
	terms_alias_organisation: ES_keyword[];
	terms_category_organisation: ES_keyword[];
	terms_mesh_description: ES_text_w_keyword[];
	terms_of_countryTags: ES_keyword[];
	terms_of_funders: ES_text_w_keyword[];
	terms_roles_organisation: ES_text_w_keyword[];
	terms_subcategory_organisation: ES_text_w_keyword[];
	textBody_descriptive_organisation: ES_text;
	textBody_summary_organisation: ES_text;
	type_of_entity: ES_keyword;
	type_of_organisation: ES_keyword;
	url_crunchBase_organisation: ES_keyword;
	url_facebook_organisation: ES_keyword;
	url_linkedIn_organisation: ES_keyword;
	url_of_organisation: ES_keyword;
	url_twitter_organisation: ES_keyword;
};

export type mosaic_crunchbase_v5 = {
	_cost_usd2018_organisation: float;
	_terms_sdg_summary: ES_keyword[];
	address_of_organisation: ES_keyword;
	booleanFlag_health_organisation: boolean;
	coordinate_of_city: ES_geo_point_obj;
	cost_of_funding: float;
	count_employee_organisation: ES_keyword;
	count_rounds_funding: integer;
	currency_of_funding: ES_keyword;
	date_birth_organisation: Date_YYYYMMDD_dash;
	date_death_organisation: Date_YYYYMMDD_dash;
	date_last_funding: Date_YYYYMMDD_dash;
	datetime_updated_organisation: Date_YYYYMMDD_dash_time;
	id_continent_organisation: ES_keyword;
	id_iso2_country: ES_keyword;
	id_iso3_country: ES_keyword;
	id_isoNumeric_country: integer;
	id_of_continent: ES_keyword;
	id_parent_organisation: ES_keyword;
	id_state_organisation: ES_keyword;
	name_of_organisation: ES_text_w_keyword;
	placeName_city_organisation: ES_text_w_keyword;
	placeName_continent_organisation: ES_text_w_keyword;
	placeName_country_organisation: ES_text_w_keyword;
	placeName_region_organisation: ES_text_w_keyword;
	placeName_state_organisation: ES_text_w_keyword;
	rank_rhodonite_organisation: float;
	status_of_organisation: ES_keyword;
	terms_alias_organisation: ES_keyword[];
	terms_category_organisation: ES_keyword[];
	terms_mesh_description: ES_text_w_keyword[];
	terms_of_countryTags: ES_keyword[];
	terms_of_funders: ES_text_w_keyword[];
	terms_roles_organisation: ES_text_w_keyword[];
	terms_subcategory_organisation: ES_text_w_keyword[];
	textBody_descriptive_organisation: ES_text_w_keyword;
	textBody_summary_organisation: ES_text_w_keyword;
	type_of_entity: ES_text_w_keyword;
	type_of_organisation: ES_keyword;
	url_crunchBase_organisation: ES_text_w_keyword;
	url_facebook_organisation: ES_text_w_keyword;
	url_linkedIn_organisation: ES_text_w_keyword;
	url_of_organisation: ES_text_w_keyword;
	url_twitter_organisation: ES_text_w_keyword;
};

export type mosaic_meetup_v3 = {
	_id_state_group: ES_keyword;
	_placeName_state_group: ES_text_w_keyword;
	_terms_memberOrigin_group: ES_text_w_keyword[];
	_terms_sdg_description: ES_keyword[];
	booleanFlag_autotranslated_entity: boolean;
	coordinate_of_group: ES_geo_point_obj;
	count_member_group: integer;
	date_start_group: Date_YYYYMMDD_dash;
	id_continent_group: ES_keyword;
	id_country_group: ES_keyword;
	id_iso2_country: ES_keyword;
	id_iso3_country: ES_keyword;
	id_isoNumeric_country: integer;
	id_of_group: ES_keyword;
	name_of_category: ES_text_w_keyword;
	name_of_group: ES_text_w_keyword;
	placeName_city_group: ES_text_w_keyword;
	placeName_continent_group: ES_text_w_keyword;
	placeName_country_group: ES_text_w_keyword;
	rank_rhodonite_group: float;
	terms_iso2lang_entity: ES_keyword[];
	terms_mesh_group: ES_text_w_keyword[];
	terms_of_countryTags: ES_keyword[];
	terms_topics_group: ES_text_w_keyword[];
	textBody_descriptive_group: ES_text;
	type_of_entity: ES_keyword;
	url_of_group: ES_keyword;
};

export type mosaic_meetup_v4 = {
	_id_state_group: ES_keyword;
	_placeName_state_group: ES_text_w_keyword;
	_terms_memberOrigin_group: ES_text_w_keyword[];
	_terms_sdg_description: ES_keyword[];
	booleanFlag_autotranslated_entity: boolean;
	coordinate_of_group: ES_geo_point_obj;
	count_member_group: integer;
	date_start_group: Date_YYYYMMDD_dash;
	id_continent_group: ES_keyword;
	id_country_group: ES_keyword;
	id_iso2_country: ES_keyword;
	id_iso3_country: ES_keyword;
	id_isoNumeric_country: integer;
	id_of_group: ES_keyword;
	name_of_category: ES_text_w_keyword;
	name_of_group: ES_text_w_keyword;
	placeName_city_group: ES_text_w_keyword;
	placeName_continent_group: ES_text_w_keyword;
	placeName_country_group: ES_text_w_keyword;
	rank_rhodonite_group: float;
	terms_iso2lang_entity: ES_keyword[];
	terms_mesh_group: ES_text_w_keyword[];
	terms_of_countryTags: ES_keyword[];
	terms_topics_group: ES_text_w_keyword[];
	textBody_descriptive_group: ES_text_w_keyword;
	type_of_entity: ES_keyword;
	url_of_group: ES_text_w_keyword;
};

export type mosaic_nih_v5 = {
	booleanFlag_duplicate_abstract: boolean;
	coordinate_of_organisation: ES_geo_point_obj;
	cost_total_project: float;
	currency_total_cost: ES_keyword;
	date_end_project: Date_YYYYMMDD_dash;
	date_start_project: Date_YYYYMMDD_dash;
	id_iso2_country: ES_keyword;
	id_iso3_country: ES_keyword;
	id_isoNumeric_country: integer;
	id_of_continent: ES_keyword;
	id_of_project: string;
	id_state_organisation: ES_keyword;
	json_funding_project: object[];
	placeName_city_organisation: ES_text_w_keyword;
	placeName_continent_organisation: ES_text_w_keyword;
	placeName_country_organisation: ES_text_w_keyword;
	placeName_state_organisation: ES_text_w_keyword;
	placeName_zipcode_organisation: ES_keyword;
	rank_rhodonite_abstract: float;
	terms_descriptive_project: ES_text[];
	terms_mesh_abstract: ES_text_w_keyword[];
	terms_of_countryTags: ES_keyword[];
	terms_of_funders: ES_text_w_keyword[];
	terms_sdg_abstract: ES_keyword[];
	textBody_abstract_project: ES_text;
	textBody_descriptive_project: ES_text;
	title_of_funder: ES_text;
	title_of_organisation: ES_text;
	title_of_project: ES_text;
	type_of_entity: ES_keyword;
	year_fiscal_funding: integer;
};

export type mosaic_nih_v6 = {
	booleanFlag_duplicate_abstract: boolean;
	coordinate_of_organisation: ES_geo_point_obj;
	cost_total_project: float;
	currency_total_cost: ES_keyword;
	date_end_project: Date_YYYYMMDD_dash;
	date_start_project: Date_YYYYMMDD_dash;
	id_iso2_country: ES_keyword;
	id_iso3_country: ES_keyword;
	id_isoNumeric_country: integer;
	id_of_continent: ES_keyword;
	id_of_project: string;
	id_state_organisation: ES_keyword;
	json_funding_project: object[];
	placeName_city_organisation: ES_text_w_keyword;
	placeName_continent_organisation: ES_text_w_keyword;
	placeName_country_organisation: ES_text_w_keyword;
	placeName_state_organisation: ES_text_w_keyword;
	placeName_zipcode_organisation: ES_keyword;
	rank_rhodonite_abstract: float;
	terms_descriptive_project: ES_text_w_keyword[];
	terms_mesh_abstract: ES_text_w_keyword[];
	terms_of_countryTags: ES_keyword[];
	terms_of_funders: ES_text_w_keyword[];
	terms_sdg_abstract: ES_keyword[];
	textBody_abstract_project: ES_text_w_keyword;
	textBody_descriptive_project: ES_text_w_keyword;
	title_of_funder: ES_text_w_keyword;
	title_of_organisation: ES_text_w_keyword;
	title_of_project: ES_text_w_keyword;
	type_of_entity: ES_keyword;
	year_fiscal_funding: integer;
};

export type mosaic_nih_v7 = {
	booleanFlag_duplicate_abstract: boolean;
	coordinate_of_organisation: ES_geo_point_obj;
	cost_total_project: float;
	currency_total_cost: ES_keyword;
	date_end_project: Date_YYYYMMDD_dash;
	date_start_project: Date_YYYYMMDD_dash;
	id_iso2_country: ES_keyword;
	id_iso3_country: ES_keyword;
	id_isoNumeric_country: integer;
	id_of_continent: ES_keyword;
	id_of_project: string;
	id_state_organisation: ES_keyword;
	json_funding_project: object[];
	placeName_city_organisation: ES_text_w_keyword;
	placeName_continent_organisation: ES_text_w_keyword;
	placeName_country_organisation: ES_text_w_keyword;
	placeName_state_organisation: ES_text_w_keyword;
	placeName_zipcode_organisation: ES_keyword;
	rank_rhodonite_abstract: float;
	terms_descriptive_project: ES_text_w_keyword[];
	terms_mesh_abstract: ES_text_w_keyword[];
	terms_of_countryTags: ES_keyword[];
	terms_of_funders: ES_text_w_keyword[];
	terms_sdg_abstract: ES_keyword[];
	textBody_abstract_project: ES_text_w_keyword;
	textBody_descriptive_project: ES_text_w_keyword;
	title_of_funder: ES_text_w_keyword;
	title_of_organisation: ES_text_w_keyword;
	title_of_project: ES_text_w_keyword;
	type_of_entity: ES_keyword;
	year_fiscal_funding: integer;
};
const selection: Aggs<eurito_crunchbase_v0, 'date_death_organisation'> = {"primary":{"avg":{field:"date_death_organisation"}}};
`;
