// elasticsearch/aggs/ref/typeGroups.js
import {
	Date_YYYYMMDD_dash_time,
	Date_YYYYMMDD_dash,
	Date_YYYYMMDD_T_dash_time_ms,
	date,
	geoPointObject,
	keyword,
	keywordArray,
	text,
	textArray,
	textWithKeyword,
	textWithKeywordArray,
} from 'app/elasticsearch/types/esTypes/fields';
import {
	boolean,
	float,
	integer,
	unionOf,
} from 'app/elasticsearch/types/genericTypes';

export const esDates = unionOf(
	date,
	Date_YYYYMMDD_dash,
	Date_YYYYMMDD_dash_time,
	Date_YYYYMMDD_T_dash_time_ms
);

export const esNumeric = unionOf(
	boolean,
	date,
	Date_YYYYMMDD_dash,
	Date_YYYYMMDD_dash_time,
	float,
	integer,
);

export const esNumericButBoolean = unionOf(
	date,
	Date_YYYYMMDD_dash,
	Date_YYYYMMDD_dash_time,
	float,
	integer,
);

export const esSearchableField = unionOf(
	boolean,
	Date_YYYYMMDD_dash_time,
	Date_YYYYMMDD_dash,
	date,
	float,
	geoPointObject,
	integer,
	keyword,
	keywordArray,
	textWithKeyword,
	textWithKeywordArray,
);

export const esSearchableString = unionOf(
	keyword,
	keywordArray,
	textWithKeyword,
	textWithKeywordArray,
);

export const esStrings = unionOf(
	keyword,
	keywordArray,
	text,
	textArray,
	textWithKeyword,
	textWithKeywordArray,
);

// TODO at some point we might need to use
// https://www.elastic.co/guide/en/elasticsearch/reference/7.9/mapping-types.html
